akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/internal/ClusterActorSetImpl.scala

```diff
@@ -0,0 +1,178 @@
+          }
+          .receiveSignal {
+            case (_, Terminated(_)) =>
+              // if the child stopped manually, but we are not being re-located to another node,
+              // we can fast-restart it instead of waiting for a ping
+              context.log.debug(s"ClusterActorSet actor ${entityContext.entityId} stopped, starting again")
+              startChild()
+              Behaviors.same
+          }
+      }
+
+    private def waitingForChildToStop(): Behavior[Command] = Behaviors.setup { context =>
+      // FIXME timeout and force termination if child doesn't stop?
+      Behaviors.receiveSignal {
+        case (_, Terminated(_)) =>
+          // if the child stopped manually, but we are not being re-located to another node,
+          // we can fast-restart it instead of waiting for a ping
+          context.log.debug(s"ClusterActorSet actor stopped")
+          Behaviors.stopped
+      }
+    }
+  }
+
+  object KeepAlivePinger {
+    sealed trait Event
+    case object Tick extends Event
+
+    def apply(
+        settings: ClusterActorSetSettings,
+        identities: Set[EntityId],
+        shardingRef: ActorRef[ShardingEnvelope[EntityParent.Command]]): Behavior[Event] =
+      Behaviors.setup { context =>
+        Behaviors.withTimers { timers =>
+          context.log.debug(
+            s"Starting ClusterActorSet KeepAlivePinger with ping interval ${settings.keepAliveInterval}")
+          // FIXME should we delay initial tick with random fraction of keepAliveInterval to avoid thundering herd?
+          timers.startTimerWithFixedDelay(Tick, settings.keepAliveInterval)
+          Behaviors.receiveMessage {
+            case Tick =>
+              identities.foreach(id => shardingRef ! ShardingEnvelope(id, EntityParent.Ping))
+              Behaviors.same
+          }
+        }
+      }
+  }
+
+}
+@InternalApi
+private[akka] class ClusterActorSetImpl(system: ActorSystem[_]) extends ClusterActorSet {
+
+  private val typeKeyCounter = new AtomicInteger(0)
+
+  override def init[T](numberOfEntities: Int, behaviorFactory: EntityId => Behavior[T], stopMessage: T): Unit =
+    init(ClusterActorSetSettings(system), numberOfEntities, behaviorFactory, stopMessage)
+
+  override def init[T](
+      settings: ClusterActorSetSettings,
+      numberOfEntities: Int,
+      behaviorFactory: EntityId => Behavior[T],
+      stopMessage: T): Unit = {
+    val identities = (0 to numberOfEntities).map(_.toString).toSet
+    init(settings, identities, behaviorFactory, stopMessage)
+  }
+
+  override def init[T](identities: Set[EntityId], behaviorFactory: EntityId => Behavior[T], stopMessage: T): Unit =
+    init(ClusterActorSetSettings(system), identities, behaviorFactory, stopMessage)
+
+  override def init[T](
+      settings: ClusterActorSetSettings,
+      identities: Set[EntityId],
+      behaviorFactory: EntityId => Behavior[T],
+      stopMessage: T): Unit = {
+    val setId = typeKeyCounter.incrementAndGet()
```

Comment:
how would this work from different nodes? I think that identifier must be given by the user

Reply:
Bad luck thinking there

Reply:
I'm thinking name and count and skip the set signature, that way we can deterministically create a set of names that will shard nicely.

Reply:
That’s not enough. The id may have business value that the actor needs, such as which tag slice read side processor should read from.

Reply:
My thinking is that if we make the entity id a `$name-$id` and pass the numeric id to the behavior factory for the actors that would be enough to chose whatever business metadata for that id it needs?

Plus we can use the id for the shard id as well.

Is that not good enough, am I missing something?

Reply:
The actors could use a fixed Map(index->businessKey) themselves, so yes it could be enough.

Nice property is that the number can be changed in a rolling update.